### PR TITLE
Yolotzin Dunbar - Time

### PR DESCRIPTION
**Pull request by ydunbar**
# Assignment Submission: Hotel

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------

What was a design challenge that you encountered on this project? | Figuring out which responsibilities/methods should be handled by which class was a challenge that required some trial and error learning. Also just deciding whether I should follow Ada's example scaffolding or going with a different plan took me a long time to think over.
-- | --
What was a design decision you made that changed over time over the project? | At first I wanted Blocks to be a child of HotelController, and then I considered making it a child of Reservations, because it would have similar instance variables and methods. In the end I decided it would be easier to treat it as a new distinct class.
What was a concept you gained clarity on, or a learning that you'd like to share? | I struggled a lot with Before blocks in my tests, because I didn't realize that it is like any other method; it's variables need to be instance variables or they won't have scope outside of the block. I was only able to see this clearly after talking with a TA who pointed it out to me.
What is an example of a nominal test that you wrote for this assignment? What makes it a nominal case? | Testing that each initializer method returned an instance of it's class had to be the most nominal test - it's creating the object it's meant to create, with the expected arguments.
What is an example of an edge case test that you wrote for this assignment? What makes it an edge case? | My DateRange test for "Raises exception for invalid input" is an edge case test, since it's unexpected that a user would enter an check-out date that is earlier than a check-in date.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I'm comfortable writing in pseudocode, because it's a quicker way for me to think through code without worrying about the exact syntax. Writing tests is still new to me and was very challenging for me to do correctly, and did not feel intuitive - I had to really think about what my tests were actually checking for.

### CODE REVIEW

File: lib/date_range.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+          @check_out = check_out
+        else
+          raise ArgumentError, "Start date must be before end date"
+        end
+      end
+      @nights = (@check_out - @check_in).to_i
+    end
+
+    def overlap?(new_date_range)
+      if (new_date_range.check_in >= self.check_out || new_date_range.check_out <= self.check_in)
```
Comment (jmaddox19) [inline lib/date_range.rb:21]: Clever!

File: lib/date_range.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        return false
+      else
+        return true
+      end
```
Comment (jmaddox19) [inline lib/date_range.rb:25]: This is very minor but this block of code could be simplified to what I have below and it would do the same thing.
```suggestion
      return !(new_date_range.check_in >= self.check_out || new_date_range.check_out <= self.check_in)
```

Comment (jmaddox19) [reply lib/date_range.rb:25]: After looking at it though, I think the way you have it may be easier to read and understand.

File: test/date_range_test.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      expect { Hotel::DateRange.new(check_in: start_date, check_out: end_date) }.must_raise ArgumentError
+    end
+  end
+
+  describe "overlap?" do
+    it "checks for schedule conflict" do
+      expect(@date_range.overlap?(@date_range)).must_equal true
+      expect(@date_range.overlap?(@other_date_range)).must_equal false
+    end
+  end
```
Comment (jmaddox19) [inline test/date_range_test.rb:26]: It'd be good to have more unit tests here for more cases. (for example all the examples we went over in class on Wednesday that week.)

**Comment by jmaddox19** [comment]
# Hotel






## Section 1: Major Learning Goals



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices SRP by having at least two separate classes with distinct responsibilities, and test files for these two classes | ✔️
Overall, demonstrates understanding instance variables vs. local variables. (There aren't unnecessarily too many instance variables, when it should be a local variable) | ✔️
For each test file, tests demonstrate an understanding of instantiating objects properly, and using Arrange-Act-Assert | ✔️
Practices pseudocode and TDD, and reflected on it by filling out the reflection questions | ✔️
Practices git with at least 15 small commits and meaningful commit messages | ✔️

## Section 2: Code Review and Testing Requirements

| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
There is a class that represents a reservation, and a second class that holds/manages a collection of reservations through composition (instance variable) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date is complex logic that is separated into method(s) (and potentially class(es)) | ✔️ well no, but only because its so clever and succinct 
The logic for checking if a reservation's date overlaps with another reservation's date has unit tests | ✔️ Yes but minimal
All of the given tests run and pass | ✔️
A test coverage tool is installed and used, and shows 95% test coverage | ✔️ 100%! 

## Section 3: Feature Requirements

| Feature Requirement: There is a method that... | yes/no |
| --- | --- |
gives back a list of rooms, and it's tested | ✔️
creates a specific reservation for a room for a given date range, and it has nominal test cases | ✔️
creates a specific reservation for a room for a given date range, and it tests an edge case, such as no available room, or invalid date range | 
gives back a list of reservations on a given date. Its tests include a test that makes several reservations for a given date | ✔️
calculates the total price for a reservation | ✔️
gives back a list of available rooms for a given date range, and it has nominal test cases | ✔️
gives back a list of available rooms for a given date range, and it has edge test cases, such as no available room, or invalid date range | 
creates a block of rooms | ✔️
reserves a room from a block | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ total in all sections | ✔️
| Yellow (Approaches Standards) | 9-13 total in all sections |
| Red (Not at Standard) | 0-8 total in all sections, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |

### Additional Feedback



Great work overall! You've built your first project with minimal starting code. This represents an incredible milestone in your journey, and you should be proud of yourself!

I am particularly impressed by the way that you setup 4 different classes that interact with each other in appropriate and interesting ways!

I do see some room for improvement around unit testing. As you noted in your reflection answers, coming up with test cases can be challenging. Though your code has 100% test coverage, there are many more opportunities for testing edge cases.


I often find it helpful to take time to think of all the tests I want to write by writing all the test names (also called "it-blocks") then going back and filling in the tests with actual code. This is similar to writing pseudocode before writing code.


I'm happy to take time to talk with you about that in person if you could use more of a conversation about testing and thinking of test cases.

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
